README.md
# Yet Another PSARC

Cross platform update of Matthieu Milan's [PSARC PS3 extractor for Linux](https://www.ferb.fr/ps3/PSARC/).  The initial commit includes minimal changes but a new command line argument was added.  *-h* tells psarc to print header information to std out without formatting the file size, using colons as field separators, and newlines to delineate entries (id:size:path). This is useful for piping information to another program that would build a tree view of an archive(s) contents.  A complete list of changes can be found in [CHANGELOG.md](CHANGELOG.md).

## Getting Started

Linux: To compile, navigate to the project directory: and execute *make* (*debug* and *clean* are also acceptable targets).

Windows:  Assuming you have MinGW and zlib headers installed, you can compile by navigating to the project directory and executing *mingw32-make.exe* (*debug* and *clean* are also acceptable targets).

Execute psarc without arguments for use instructions.

### How to Setup MinGW

For a verbose and illustrated guide visit [codebind](http://www.codebind.com/cprogramming/install-mingw-windows-10-gcc/).  For those who prefer brevity, here is the tl;dr:

Download and Install 
* Download the MinGW setup [here](https://osdn.net/projects/mingw/downloads/68260/mingw-get-setup.exe).
* Launch the setup and click on *Install*.
* Accept the default options and click *Continue* twice.
* Once the package manager is up click on *Basic Setup* in the left pane
* In the right pane, right click *mingw32-base-bin* and click *Mark for Installation*
* In the right pane, right click *mingw32-gcc-g++-bin* and click *Mark for Installation*
* In the left pane, click All Packages -> MinGW Libraries -> MinGW Standard Libraries
* In the right pane scroll down until you find *mingw32-libz-dev* and right click -> Mark for Installation
* From the top menu, click Installation -> Apply Changes
* In the dialogue box that pops up click *Apply* and then *Close* once it completes, then close the installation manager

Add C:\MinGW\bin to your path
* Click on the Windows start menu and type *env*, then click on "Edit the system environment variables" from the list above.
* Near the bottom of the dialogue box that pops up, click the *Environment Variables* button.
* In the top list, click *Path* and then click the *Edit* directly below that box (middle of dialogue box).
* On the right, click *New*
* Assuming you didn't change the default install location, type in *C:\MinGW\bin* and click *OK*
* Click *OK* two more times to close out the previous dialogues.

If you ever need to update or change installed packages run the MinGW shortcut on your desktop or in the start menu.  You can also launch mingw32-get from the command line or Powershell.

Congratulations, you can now open a command prompt or Powershell and compile.

## Special Notes

For those interested, I have made copies of my VS Code json files available in the *json* folder.  To use these settings copy the *.vscode* folder appropriate for your OS to the project base directory.

## Deployment

As a Windows user, if you have MinGW installed it is not necessary to statically link libraries and the LDFLAGS variables can be commented out of the Makefile (# is used for comments).  This will reduce your executable size from over 400kb to around 92kb.

To comment out static linking change this:
```
LDFLAGS = -static-libstdc++ -static-libgcc -static
```
To This:
```
#LDFLAGS = -static-libstdc++ -static-libgcc -static
```

## Built With

* [Visual Studio Code](https://github.com/Microsoft/vscode) - IDE used for both Windows and Linux
* [MinGW](http://mingw.org/) - Windows development environment

## Authors

* **Paul Hedrick** - *Rework* - [github](https://github.com/paul-hedrick) & [website](https://phedrick.dev) - not currently active

## Acknowledgments

* Matthieu Milan for the original source


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.1.3.1] - 2019-10-06
### Added
- Added *-h* command to print header information to std out (id:size:full/file/path).

### Changed
- Complete rewrite of Makefile to support Windows and add *debug* target.
- Changed *mkdir* to a macro to support cross platform directory manipulation.
- Windows: Added include for *direct.h* and replacement define for *fseeko* to *fseek*.
- Tweaked formatting of usage message.
- Moved *_buffer* size and temp file path to defines.
- Windows: Added basic implementation of strndup and PRId64 which are missing from MinGW.
- Changed *inflateEntry*'s file open (write) to binary mode to prevent fseek from choking on carriage returns.
- Optimized how file paths are read into entires.


src/sys.h
/*
 * Open PSARC PS3 extractor
 * Copyright (C) 2011-2018 Matthieu Milan
 */

#ifndef SYS_H__
#define SYS_H__

#define _FILE_OFFSET_BITS 64

#include <errno.h>
#include <libgen.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <zlib.h>

#ifndef __linux__
	#include <direct.h>  //needed for mkdir in windows
	#define fseeko fseek //no fseeko in mingw, fseek is not usable with text files
	#define __mkdir__(p,m) mkdir(p) //windows doesn't have perms
#else
	#define __mkdir__(p,m) mkdir(p,m) //and linux does
#endif

inline uint16_t READ_BE_UINT16(const uint8_t *ptr) {
	return (ptr[0] << 8) | ptr[1];
}

inline uint32_t READ_BE_INT24(const uint8_t *ptr) {
	return (ptr[0] << 16) | (ptr[1] << 8) | ptr[2];
}

inline uint32_t READ_BE_UINT32(const uint8_t *ptr) {
	return (ptr[0] << 24) | (ptr[1] << 16) | (ptr[2] << 8) | ptr[3];
}

inline uint64_t READ_BE_INT40(const uint8_t *ptr) {
	uint64_t res = ptr[4] | (0x100L * ptr[3]) | (ptr[2] * 0x10000L) | (ptr[1] * 0x1000000L) | (ptr[0] * 0x100000000LLU);
	if (ptr[0] == 0)
		res %= 0xffffffff00000000LLU;
	else
		res %= 0xfffffffe00000000LLU;
	return res;
}

inline int mkpath(const char *s, mode_t mode){
	char *q, *r = NULL, *path = NULL, *up = NULL;
	int rv;

	rv = -1;
	if (strcmp(s, ".") == 0 || strcmp(s, "/") == 0)
		return (0);

	if ((path = strdup(s)) == NULL)
		exit(1);

	if ((q = strdup(s)) == NULL)
		exit(1);

	if ((r = dirname(q)) == NULL) {
		free(q);
		free(path);
		return (rv);
	}

	if ((up = strdup(r)) == NULL)
		exit(1);

	if ((mkpath(up, mode) == -1) && (errno != EEXIST)) {
		free(up);
		free(q);
		free(path);
		return (rv);
	}

	if ((__mkdir__(path, mode) == -1) && (errno != EEXIST))
		rv = -1;
	else
		rv = 0;

	free(up);
	free(q);
	free(path);
	return (rv);
}

#endif // SYS_H__


src/psarc.h
/*
 * Open PSARC PS3 extractor
 * Copyright (C) 2011-2018 Matthieu Milan
 */

#ifndef PSARC_H__
#define PSARC_H__

#include "file.h"

#define kPSARCMagicNumber 0x50534152

#define BUFFER_SIZE 600*1024				//added defines for a few literals
#define TEMP_FILE "psarc.temp"				//that I had to modify
#ifndef __linux__							//getting windows temp path requires
	#define TEMP_DIR  "."					//filesystem.h which isn't an option
	#define TEMP_PATH "./psarc.temp"		//using working directory for temp file
	#define PRId64 "lld"
#else
	#define TEMP_DIR  "/tmp"
	#define TEMP_PATH "/tmp/psarc.temp"
#endif


struct Pack {
	int32_t _id;
	uint64_t _length;
	char *_name;
	uint32_t _zIndex;
	uint64_t _zOffset;
};

struct PSARC {
	File _f;
	uint8_t *_buffer;

	Pack *_entries;
	uint32_t _numEntries;
	uint32_t _maxLen;

	PSARC();
	~PSARC();

	void inflateEntry(uint32_t entry, uint32_t *zBlocks, uint32_t cBlockSize, char *basename, char *dirname);
	void read(const char *arcName, uint32_t start, uint32_t end, bool header, bool console = false);
	void read(const char *arcName, uint32_t start, uint32_t end);
	void readHeader(const char *arcName, bool console = false);
};

#ifndef __linux__
	char * strndup (const char *s, size_t n); //not implemented in mingw
#endif

#endif // PSARC_H__


src/file.cpp
/*
 * Open PSARC PS3 extractor
 * Copyright (C) 2011-2018 Matthieu Milan
 */

#include "file.h"

struct File_impl {
	bool _ioErr;
	uint64_t _offset;
	File_impl() : _ioErr(false) {}
	virtual ~File_impl() {}
	virtual bool open(const char *path, const char *mode) = 0;
	virtual void close() = 0;
	virtual uint64_t offset() = 0;
	virtual void seek(uint64_t off) = 0;
	virtual void shift(uint64_t off) = 0;
	virtual void read(void *ptr, uint32_t size) = 0;
	virtual void write(void *ptr, uint32_t size) = 0;
};

struct stdFile : File_impl {
	FILE *_fp;
	stdFile() : _fp(0) {}
	bool open(const char *path, const char *mode) {
		_ioErr = false;
		_offset = 0;
		_fp = fopen(path, mode);
		return (_fp != 0);
	}
	void close() {
		if (_fp) {
			fclose(_fp);
			_fp = 0;
		}
	}
	void seek(uint64_t off) {
		if (_fp) {
			_offset = off;
			fseeko(_fp, off, SEEK_SET);
		}
	}
	void shift(uint64_t off) {
		if (_fp) {
			_offset += off;
			fseeko(_fp, _offset, SEEK_SET);
		}
	}
	uint64_t offset() {
		if (_fp) {
			return _offset;
		}

		return 0; // TODO: trouver mieux
	}
	void read(void *ptr, uint32_t size) {
		if (_fp) {
			_offset += size;
			uint32_t r = fread(ptr, 1, size, _fp);
			if (r != size)
				_ioErr = true;
		}
	}
	void write(void *ptr, uint32_t size) {
		if (_fp) {
			_offset += size;
			uint32_t r = fwrite(ptr, 1, size, _fp);
			if (r != size)
				_ioErr = true;
		}
	}
};

File::File() {
	_impl = new stdFile;
}

File::~File() {
	if (_impl) {
		_impl->close();
		delete _impl;
	}
}

bool File::open(const char *filename, const char *directory, const char *mode) {
	if (_impl) {
		_impl->close();
		delete _impl;
		_impl = 0;
	}

	if (!_impl)
		_impl = new stdFile;

	char buf[512];
	snprintf(buf, sizeof(buf), "%s/%s", directory, filename);

	return _impl->open(buf, mode);
}

void File::close() {
	if (_impl) {
		_impl->close();
	}
}

bool File::ioErr() const {
	return _impl->_ioErr;
}

void File::seek(uint64_t off) {
	_impl->seek(off);
}

void File::shift(uint64_t off) {
	_impl->shift(off);
}

uint64_t File::offset() {
	return _impl->offset();
}

void File::read(void *ptr, uint32_t size) {
	_impl->read(ptr, size);
}

uint8_t File::readByte() {
	uint8_t b;
	read(&b, 1);

	return b;
}

uint32_t File::readInt24BE(uint8_t *ptr) {
	read(ptr, 3);

	return READ_BE_INT24(ptr);
}

uint64_t File::readInt40BE(uint8_t *ptr) {
	read(ptr, 5);

	return READ_BE_INT40(ptr);
}

uint16_t File::readUint16BE(uint8_t *ptr) {
	read(ptr, 2);

	return READ_BE_UINT16(ptr);
}

uint32_t File::readUint32BE(uint8_t *ptr) {
	read(ptr, 4);

	return READ_BE_UINT32(ptr);
}

void File::write(void *ptr, uint32_t size) {
	_impl->write(ptr, size);
}


src/main.cpp
/*
 * Open PSARC PS3 extractor
 * Copyright (C) 2011-2018 Matthieu Milan
 */

#include "psarc.h"

void usage() {
	printf("Usage: psarc [option] filename\n");
	printf("Options:\n");
	printf("\t-h\t\tPrint colon seperated header information to std out (id:size:name)\n");
	printf("\t-l\t\tCreate a text file that lists the file id, size, and name of\n\t\t\tevery file in the archive.\n");
	printf("\t-x\t\tExtracts all files.\n");
	printf("\t-e START END\tExtracts files with the file id specified by the range between\n\t\t\tSTART and END (inclusive).\n");
	printf("\t-v\t\tDisplay version.\n");
}

int main(int argc, char *argv[]) {
	PSARC psarc;

	switch (argc) {
		case 2:
			{
				if (strncmp(argv[1], "-v", 2) == 0)
					printf("psarc: version 0.1.3.1\n");
				else
					usage();
			}
			break;

		case 3:
			{
				const char *arcName = argv[2];
				if (strncmp(argv[1], "-h", 2) == 0)
					psarc.readHeader(arcName, true);
				else if (strncmp(argv[1], "-l", 2) == 0)
					psarc.readHeader(arcName);
				else if (strncmp(argv[1], "-x", 2) == 0)
					psarc.read(arcName, 0, 0);
			}
			break;

		case 5:
			{
				if (strncmp(argv[1], "-e", 2) == 0) {
					int argv1 = atoi(argv[2]);
					int argv2 = atoi(argv[3]);
					if (argv1 > 0 && argv2 >= argv1)
						psarc.read(argv[4], argv1, argv2);
				}
			}
			break;

		default:
			usage();
			break;
	}

	return EXIT_SUCCESS;
}


src/psarc.cpp
/*
 * Open PSARC PS3 extractor
 * Copyright (C) 2011-2018 Matthieu Milan
 */

#include <inttypes.h>
#include "psarc.h"

PSARC::PSARC() {
	_buffer = (uint8_t *)malloc(BUFFER_SIZE);  //moved literal to define fo use in loop logic
	_entries = NULL;
}

PSARC::~PSARC() {
	_f.close();
	free(_buffer);

	if (_entries != NULL)
		delete _entries;
}

void PSARC::inflateEntry(uint32_t entry, uint32_t *zBlocks, uint32_t cBlockSize, char *basename, char *dirname) {
	uint64_t length = 0;

	if (entry == 0) {
		dirname = (char *)TEMP_DIR;
		basename = (char *)TEMP_FILE;
	}

	File stream;
	if (stream.open(basename, dirname, "wb+")) { //changed to binary mode to avoid issues with carriage returns
		if (_entries[entry]._length != 0) {
			_f.seek(_entries[entry]._zOffset);
			uint32_t zIndex = _entries[entry]._zIndex;
			do {
				if (zBlocks[zIndex] == 0) {
					_f.read(_buffer, cBlockSize);
					stream.write(_buffer, cBlockSize);
				} else {
					uint16_t isGzipped = _f.readUint16BE(_buffer);
					_f.shift(-2);
					_f.read(_buffer, zBlocks[zIndex]);
					if (isGzipped == 0x78da) {
						uint8_t *uncompressData;
						uLongf uncompressSize;
						uint32_t val = _entries[entry]._length - (zIndex - _entries[entry]._zIndex) * cBlockSize;
						if (val < cBlockSize) {
							uncompressData = (uint8_t *)malloc(val);
							uncompressSize = (uLongf)val;
						} else {
							uncompressData = (uint8_t *)malloc(cBlockSize);
							uncompressSize = (uLongf)cBlockSize;
						}
						uncompress(uncompressData, &uncompressSize, _buffer, zBlocks[zIndex]);
						stream.write(uncompressData, uncompressSize);
						free(uncompressData);
					} else
						stream.write(_buffer, zBlocks[zIndex]);
				}
				zIndex++;
			} while (stream.offset() < _entries[entry]._length);
		}

		if (entry == 0) {
			length = stream.offset();
			stream.close();

			File reader;
			reader.open(basename, dirname, "r");

			_entries[0]._name = (char *)"NamesBlock.bin";
			for (uint32_t i = 1; i < _numEntries; i++) {	//This read loop was reworked to not use fseeko
				uint8_t count = 0;							//fseeko is not imnplemented on minggw and fseek
															//chokes on text files.  This method is faster anyway.
				_buffer[count] = reader.readByte();
				while ((_buffer[count] != 10) && (reader.offset() < length) && count < (BUFFER_SIZE-1)) {
					count++;
					_buffer[count] = reader.readByte();	
				}
				uint8_t plength = (_buffer[count] == 10) ? count : (count + 1); //EOL = count, EOF = count +1
				_entries[i]._name = strndup((char *)_buffer, plength);
			}

			reader.close();
			remove(TEMP_PATH);
		} else {
			if (stream.offset() != _entries[entry]._length)
				printf("File size : %" PRId64 " bytes. Expected size: %" PRId64 " bytes\n", stream.offset(), _entries[entry]._length);

			stream.close();
		}
	}
}

void PSARC::read(const char *arcName, uint32_t start, uint32_t end, bool header, bool console) {
	char *dirNamec = strdup(arcName);
	char *fileNamec = strdup(arcName);

	char *dirName = dirname(dirNamec);
	char *fileName = basename(fileNamec);

	if (_f.open(fileName, dirName)) {
		if (_f.readUint32BE(_buffer) == kPSARCMagicNumber) {
			_f.seek(8);
			if (_f.readUint32BE(_buffer) == 0x7a6c6962) {
				uint32_t zSize = _f.readUint32BE(_buffer);
				_f.seek(20);
				_numEntries = _f.readUint32BE(_buffer);
				_f.seek(24);
				uint32_t cBlockSize = _f.readUint32BE(_buffer);

				uint8_t zType = 1;
				uint32_t i = 256;
				do {
					i *= 256;
					zType = (uint8_t)(zType + 1);
				} while (i < cBlockSize);

				_f.seek(32);
				_entries = new Pack[_numEntries];
				for (uint32_t i = 0; i < _numEntries; i++) {
					_f.shift(16);

					_entries[i]._id = i;
					_entries[i]._zIndex = _f.readUint32BE(_buffer);
					_entries[i]._length = _f.readInt40BE(_buffer);
					_entries[i]._zOffset = _f.readInt40BE(_buffer);
					_maxLen = _maxLen >= _entries[i]._length ? _maxLen : _entries[i]._length;
				}

				uint32_t numBlocks = (zSize - (uint32_t)_f.offset()) / zType;
				uint32_t *zBlocks = new uint32_t[numBlocks];
				for (uint32_t i = 0; i < numBlocks; i++) {
					switch (zType) {
						case 2:
							zBlocks[i] = _f.readUint16BE(_buffer);
							break;

						case 3:
							zBlocks[i] = _f.readInt24BE(_buffer);
							break;

						case 4:
							zBlocks[i] = _f.readUint32BE(_buffer);
							break;
					}
				}

				char *baseDir = NULL;
				char ext[] = ".psarc";
				if (strlen(fileName) >= 6 && strncmp(fileName + strlen(fileName) - strlen(ext), ext, strlen(ext)) == 0) {
					baseDir = (char *)malloc(strlen(fileName) - strlen(ext) + 1);
					snprintf(baseDir, strlen(fileName) - strlen(ext) + 1, "%s", fileName);
				} else {
					char data[] = "_data";

					baseDir = (char *)malloc(strlen(fileName) + strlen(data) + 1);
					snprintf(baseDir, strlen(fileName) + strlen(data) + 1, "%s%s", fileName, data);
				}

				inflateEntry(0, zBlocks, cBlockSize, NULL, NULL);

				if (header) {
					char ext[] = ".txt";
					char *outList = (char *)malloc(strlen(fileName) + strlen(ext) + 1);
					snprintf(outList, strlen(fileName) + strlen(ext) + 1, "%s%s", fileName, ext);

					if (console) {
						for (uint32_t i = 1; i < _numEntries; i++)
							printf("%d:%" PRId64 ":%s\n", i, _entries[i]._length, _entries[i]._name);
					} else {
						File list;
						if (list.open(outList, ".", "w")) {
							char * msg = (char *) malloc(_maxLen +1);
							for (uint32_t i = 1; i < _numEntries; i++) {
								if (_entries[i]._length < 0x400L)
									sprintf(msg, "%d %" PRId64 " b %s\n", i, _entries[i]._length, _entries[i]._name);
								else if (_entries[i]._length < 0x100000L)
									sprintf(msg, "%d %1.2f Kb %s\n", i, _entries[i]._length / 1024.0, _entries[i]._name);
								else if (_entries[i]._length < 0x40000000L)
									sprintf(msg, "%d %1.2f Mb %s\n", i, _entries[i]._length / 1048576.0, _entries[i]._name);
								else
									sprintf(msg, "%d %1.2f Gb %s\n", i, _entries[i]._length / 1073741824.0, _entries[i]._name);

								list.write(msg, strlen(msg));
							}
							free(msg);
							list.close();
						}
					}
					free(outList);
				} else {
					bool flag = true;
					if (start == 0) {
						start = 1;
						end = _numEntries;
					} else if ((start > (_numEntries - 1)) || (end > (_numEntries - 1))) {
						flag = false;
					} else {
						end++;
					}

					if (flag) {
						__mkdir__(baseDir, 0777);
						
						for (uint32_t i = start; i < end; i++) {
							printf("%i %s\n", _entries[i]._id, _entries[i]._name);

							char *subOutDirc = strdup(_entries[i]._name);
							char *outFilec = strdup(_entries[i]._name);

							char *subOutDir = dirname(subOutDirc);
							char *outFile = basename(outFilec);
							char *outDir;
							if (strncmp("/", _entries[i]._name, 1) == 0) {
								outDir = (char *)malloc(strlen(baseDir) + strlen(subOutDir) + 1);
								snprintf(outDir, strlen(baseDir) + strlen(subOutDir) + 1, "%s%s", baseDir, subOutDir);
							} else {
								outDir = (char *)malloc(strlen(baseDir) + strlen(subOutDir) + 2);
								snprintf(outDir, strlen(baseDir) + strlen(subOutDir) + 2, "%s/%s", baseDir, subOutDir);
							}

							mkpath(outDir, 0777);

							inflateEntry(i, zBlocks, cBlockSize, outFile, outDir);

							free(outDir);
							free(subOutDirc);
							free(outFilec);
							free(_entries[i]._name);
						}
					}
				}

				free(baseDir);
				delete zBlocks;
			}
		} else
			printf("Compression type is not zlib... Aborting.");
	}

	free(dirNamec);
	free(fileNamec);
}

void PSARC::read(const char *arcName, uint32_t start, uint32_t end) {
	read(arcName, start, end, false);
}

void PSARC::readHeader(const char *arcName, bool console) {
	read(arcName, 0, 0, true, console);
}

#ifndef __linux__
	char * strndup (const char *s, size_t n)
	{
		char *cpy = (char *)malloc(n + 1);
		memcpy(cpy, s, n);
		cpy[n] = '\0';
		return cpy;
	}
#endif
